### PR TITLE
recycle items in cascading

### PR DESCRIPTION
**Pull request by developerllazy**
# Please Note (you may remove this section before opening your PR):
We receive lots of PRs and it is hard to give proper review to PRs. Please make it easy on us by following these guidelines:

1. We do not accept changes to `master`. Please make sure your pull request is aimed at `dev`.
2. If you changed a bunch of files (that aren't config files) or multiple workers to implement your feature, it probably won't get proper attention. Please split it up into multiple, smaller, more focused, and iterative PRs if you can.
3. If you are adding a config value to something, make sure you update the appropriate `config.json` example files.

## Short Description:
Enhance the recycle function so that it will consider the total number in a category. If set to keep the maximum of balls as 100. When the total balls excess 100, it will recycle Pokeball, then Greatball, then Ultraball. This function can apply to pokeball, berries, potions & revives

## Fixes (provide links to github issues if you can):
-
-
-

### CODE REVIEW

File: pokemongo_bot/cell_workers/recycle_items.py [hunk 1]
```diff
@@ -67,47 +76,122 @@ def should_run(self):
         :return: True if the recycling process should be run; otherwise, False.
         :rtype: bool
         """
-        if inventory.Items.get_space_left() < (DEFAULT_MIN_EMPTY_SPACE if self.min_empty_space is None else self.min_empty_space):
+        if inventory.items().get_space_left() < (DEFAULT_MIN_EMPTY_SPACE if self.min_empty_space is None else self.min_empty_space):
```
Comment (BriceSD) [inline pokemongo_bot/cell_workers/recycle_items.py]: get_space_left() is a class method, thus it should be called by Items an not items(), please revert

**Comment by BriceSD** [comment]
@developerllazy You reverted the latest changes that have been made here : https://github.com/PokemonGoF/PokemonGo-Bot/pull/3921/files
Could you add them again please ?

**Comment by TheSavior** [comment]
Similar PR trying to solve the same thing: #3135

**Comment by BriceSD** [comment]
Given we have no news from the other PR author and that @developerllazy is using the latest recycle item task and has even added documentation, I definitely think we should prioritize this one

**Comment by developerllazy** [comment]
@BriceSD reverted back the changes made in https://github.com/PokemonGoF/PokemonGo-Bot/pull/3921/files

File: pokemongo_bot/cell_workers/recycle_items.py [hunk 2]
```diff
@@ -67,7 +76,7 @@ def should_run(self):
...
```
Comment (BriceSD) [inline pokemongo_bot/cell_workers/recycle_items.py]: Good job on reverting changes, you forgot this one though.

Comment (developerllazy) [inline pokemongo_bot/cell_workers/recycle_items.py]: Updated

**Comment by BriceSD** [comment]
:+1:

File: pokemongo_bot/cell_workers/recycle_items.py [hunk 3]
```diff
@@ -77,23 +86,97 @@ def work(self):
...
         worker_result = WorkerResult.SUCCESS
+
         if self.should_run():
 
+            if not (self.max_balls_keep is None):
+                self.recycle_excess_category_max(self.max_balls_keep, [1,2,3,4])
+            if not (self.max_potions_keep is None):
+                self.recycle_excess_category_max(self.max_potions_keep, [101,102,103,104])
+            if not (self.max_berries_keep is None):
+                self.recycle_excess_category_max(self.max_potions_keep, [701,702,703,704])
```
Comment (BriceSD) [inline pokemongo_bot/cell_workers/recycle_items.py]: Copypasta striked again  :P 
Using ```self.max_potions_keep``` 3 times

**Comment by douglascamata** [comment]
👍

**Comment by BriceSD** [comment]
see https://github.com/developerllazy/PokemonGo-Bot/pull/1/files

**Comment by BriceSD** [comment]
:+1:

**Comment by BriceSD** [comment]
@developerllazy My bad sorry I went too fast. Their is merging issues

**Comment by developerllazy** [comment]
@BriceSD Sorry, I am new to github, do I need to do anything?

**Comment by BriceSD** [comment]
@developerllazy Sorry about that, it’s on me. I didn’t think it would fail integrations tests since latest changes had no impacts over other parts of the bot. But by merging it too quickly it seems that the testing tool couldn’t find the PR to test, making all of them fail.

Making a new PR is one way to take tests again. Once I see them green I’ll merge

**Comment by developerllazy** [comment]
Ok, no problem

**Pull request merged**